src/decoder/plugins/OpusDecoderPlugin.cxx

```diff
@@ -327,6 +327,10 @@ MPDOpusDecoder::HandleAudio(const ogg_packet &packet)
 			return;
 	}
 
+	// the size of the opus packet in bits / number of samples * samp/sec / 1000
+	int channels = opus_packet_get_nb_channels((const unsigned char*)packet.packet);
+	uint16_t kbits = (unsigned int)packet.bytes*channels*8 * opus_sample_rate / (nframes * frame_size / 2) / 1000;
```

Comment:
I ... don't understand this formula.
`packet.bytes*channels*8` is supposed the "size of the opus packet in bits", right? But ... why multiply with the number of channels?
I think you wrote `/1000` because MPD expects kilobits/s and not bits/s, right?
What is `nframes * frame_size / 2`? `nframes * frame_size` would be the uncompressed size of the `opus_decode()` result, but why do you need that? And what does the `/ 2` mean?
I'm very confused.

Reply:
My bad, I could have done a better job documenting how this works. The basic idea of what this is trying to do is: (bits of input) \* (samples / second) / samples, yielding bits per second, which we divide by 1000 for kbits/second.

> But ... why multiply with the number of channels?

In theory this is supposed to be the number of bits of input that resulted in the corresponding decoded output. I'm honestly not sure why multiplying by the number of channels was necessary, but it was. I originally tried without this, but found that stereo files would report half their correct bitrate. With adding this, testing with both mono and stereo files yielded the correct results, so it seems to be necessary. There may be something weird about how ogg packages multiple channels or how the calling code reports the size.

> I think you wrote /1000 because MPD expects kilobits/s and not bits/s, right?

Yes.

> What is `nframes * frame_size / 2`?

This is the number of samples in the decoded audio. Given the size in bytes of the decoded audio, dividing by 2 gives the number of samples in 16-bit PCM since 2 bytes per sample.

Reply:
Sorry, I still don't understand your explanation. None of this appears to make any sense. And since you're "honestly not sure" how this works, can't explain it, and did random things until it appeared to work, I don't have much trust in the correctness of your code.
To calculate the bit rate, you don't need the uncompressed size **at all**. If you need it, then your formula is wrong. It's as simple as that. Your formula calculates the uncompressed size, so it's wrong.
The bit rate formula is "number_of_bits_in_input_file / decoded_time". The output frame/sample size is nowhere to be found in this formula, don't you see? Why does your formula have it?

Reply:
> And since you're "honestly not sure" how this works, can't explain it, and did random things until it appeared to work, I don't have much trust in the correctness of your code.

I have now figured out exactly why. Disregard what I said before about ogg packets. It's the output number of samples needs to be divided by the number of channels in order to obtain the correct time (since the two channels are played simultaneously), not the input that needs to be multiplied. It just turned out that multiplying the input size by this was mathematically equivalent to dividing the output size. I will push a new commit moving this value to the correct spot and adding better documentation. Sorry about the confusion.

Also, I have been testing this on many opus files for several days now and it's accurate as far as I can determine by comparison with what `mpv` reports, including reporting the exact intended bitrate of CBR files.

I'll explain the formula step-by-step, and with the units of the calculation resulting from each step

```
bits: number_of_bits_before_decoding = packet.bytes*8
bytes: number_of_bytes_of_audio_decoded = nframes * frame_size
samples: number_of_samples_of_audio_decoded_per_channel = number_of_bytes_of_audio_decoded / channels / 2     // because 16-bit PCM audio by definition contains exactly two bytes per sample.
1/seconds: reciprocal_of_decoded_time = opus_sample_rate / number_of_samples_of_audio_decoded
bits/seconds: bps = number_of_bits_before_decoding * reciprocal_of_seconds_of_audio_decoded
kilobits/seconds: kbits = bps / 1000
```

> The bit rate formula is "number_of_bits_in_input_file / decoded_time". The output frame/sample size is nowhere to be found in this formula, don't you see? Why does your formula have it?

Yes, of course, and that is exactly what I'm doing here. But we must calculate the decoded_time. We can't just pull this number out of thin air, we get it by analyzing the output. Also, the formula can also be number_of_bits \* 1 / decoded_time, as these are mathematically equivalent. That is the calculation being performed here.

>  To calculate the bit rate, you don't need the uncompressed size at all. If you need it, then your formula is wrong. It's as simple as that. Your formula calculates the uncompressed size, so it's wrong.

Again, the uncompressed size is used to calculate the decoded_time, which in turn is used to calculate the bit rate; it is not the bit rate itself. We can calculate the length in seconds of a segment of PCM audio based on the number of samples it contains and the sample rate. PCM audio provided by the opus decoder contains 48000 samples per second. This is constant. Each sample contains 2 bytes.

Reply:
> Again, the uncompressed size is used to calculate the decoded_time

But you don't need it. You used it, but you don't need to use it, to make your formula much more complicated than necessary.

Reply:
Can you suggest where else I could obtain the decoded_time? Perhaps I'm missing something.

Reply:
`opus_sample_rate / nframes` - that's all.